File: python/graph_based_algorithms/depth_first_search.py
# Define possible moves (left, right, down, up)
directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
non_traversable = {'-', '|'}

def is_traversable(grid, visited, rows, cols, x, y):
        """Check if a cell is within bounds, traversable, and not yet visited."""
        return 0 <= x < rows and 0 <= y < cols and grid[x][y] not in non_traversable and (x, y) not in visited

def dfs(grid, start, end):
    """
    Perform depth-first search (DFS) on a 2D grid to find the shortest path from start to end.

    Args:
        grid (List[List[str]]): The 2D grid representing the map.
        start (Tuple[int, int]): Coordinates of the starting point.
        end (Tuple[int, int]): Coordinates of the ending point.

    Returns:
        Tuple[List[List[str]], int]: A tuple containing:
            - The grid with the path marked.
            - The number of steps in the shortest path, or -1 if the end is unreachable.
    """
    # Initialize visited set and came_from dictionary
    visited = set()
    came_from = {}

    # Clone the grid to avoid modifying the original during path marking
    path_grid = [row[:] for row in grid]

    rows, cols = len(grid), len(grid[0])

    def dfs_recursive(x, y):
        if (x, y) == end:
            return True

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if is_traversable(grid, visited, rows, cols, nx, ny):
                visited.add((nx, ny))
                came_from[(nx, ny)] = (x, y)  # Record the previous cell
                if dfs_recursive(nx, ny):
                    return True
        return False

    # Start the DFS from the start position
    visited.add(start)
    if dfs_recursive(start[0], start[1]):
        # Backtrack to mark the path
        current = end
        while current != start:
            cx, cy = current
            path_grid[cx][cy] = '*'  # Mark each cell along the path
            current = came_from[current]
        
        # Calculate the distance (number of steps) from the start to the end
        dist = 0
        current = end
        while current != start:
            dist += 1
            current = came_from[current]
        
        return path_grid, dist
    
    # If end is unreachable
    return path_grid, -1

File: C++/graph_based_algorithms/src/depth_first_search.cpp
#include "pathfinding_utils.h"

// Recursive helper function
bool dfs_recursive(std::vector<std::vector<char>>& grid, std::pair<int, int> current, 
                   const std::pair<int, int>& end, std::set<std::pair<int, int>>& visited,
                   std::map<std::pair<int, int>, std::pair<int, int>>& came_from) 
{
    if (current == end) {
        return true; // Found the end position
    }

    for (const std::pair<int, int>& direction : directions) {
        
        int x = current.first + direction.first;
        int y = current.second + direction.second;
        std::pair<int, int> next = {x, y};

        // Check boundaries and traversability
        if (isTraversable(grid, visited, x, y)) 
        {
            visited.insert(next);
            came_from[next] = current;

            // Recursive call
            if (dfs_recursive(grid, next, end, visited, came_from)) {
                return true;
            }
        }
    }
    return false;
}

int dfs(std::vector<std::vector<char>>& grid) 
{
    // get start and end positions
    std::pair<int, int> start, end;
    std::tie(start, end) = findLocations(grid);
    
    // Initialize the stack
    std::stack<std::pair<std::pair<int, int>, int>> stack;
    stack.push({start, 0});

    // Track visited cells
    std::set<std::pair<int, int>> visited; 
    visited.insert(start);

    std::map<std::pair<int, int>, std::pair<int, int>> came_from; // slower than unsing std::unordered_map
    came_from[start] = {-1, -1};
    
    // Call the recursive DFS function
    if (dfs_recursive(grid, start, end, visited, came_from)) {
        // Backtrack to mark the path
        std::pair<int, int> current = end;
        int distance = 0;

        while (current != start) {
            grid[current.first][current.second] = '*'; // Mark the path
            current = came_from[current];
            ++distance;
        }
        return distance; // Return the path length
    }

    return -1; // End is unreachable
}